modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java

```diff
@@ -235,4 +288,209 @@ public static boolean recoverable(Throwable t) {
+                            ops().yield()
+                        )
+        );
+
+        return metaStorageMgr.invoke(iif).thenApply(unused -> null);
+    }
+
+    /**
+     * Builds a list of cluster nodes based on a list of peers, pending and stable assignments.
+     * A peer will be added to the result list iff peer's address is present in pending or stable assignments.
+     *
+     * @param peers List of peers.
+     * @param pendingAssignments Byte array that contains serialized list of pending assignments.
+     * @param stableAssignments Byte array that contains serialized list of stable assignments.
+     * @return Resolved cluster nodes.
+     */
+    public static List<ClusterNode> resolveClusterNodes(List<PeerId> peers, byte[] pendingAssignments, byte[] stableAssignments) {
+        Map<NetworkAddress, ClusterNode> resolveRegistry = new HashMap<>();
+
+        if (pendingAssignments != null) {
+            List<ClusterNode> pending = ByteUtils.fromBytes(pendingAssignments);
+            pending.forEach(n -> resolveRegistry.put(n.address(), n));
+        }
+
+        if (stableAssignments != null) {
+            List<ClusterNode> stable = ByteUtils.fromBytes(stableAssignments);
+            stable.forEach(n -> resolveRegistry.put(n.address(), n));
+        }
+
+        List<ClusterNode> resolvedNodes = new ArrayList<>(peers.size());
+
+        for (PeerId p : peers) {
+            var addr = NetworkAddress.from(p.getEndpoint().getIp() + ":" + p.getEndpoint().getPort());
+
+            if (resolveRegistry.containsKey(addr)) {
+                resolvedNodes.add(resolveRegistry.get(addr));
+            } else {
+                throw new IgniteInternalException("Can't find appropriate cluster node for raft group peer: " + p);
+            }
+        }
+
+        return resolvedNodes;
+    }
+
+    /**
+     * Reads a list of cluster nodes from a MetaStorage entry.
+     *
+     * @param entry MetaStorage entry.
+     * @return List of cluster nodes.
+     */
+    public static List<ClusterNode> readClusterNodes(Entry entry) {
+        if (entry.empty()) {
+            return Collections.emptyList();
+        }
+
+        return ByteUtils.fromBytes(entry.value());
+    }
+
+    /**
+     * Removes nodes from collection of nodes.
+     *
+     * @param minuend Collection to remove nodes from.
+     * @param subtrahend Collection of nodes to be removed.
+     * @return Result of the subtraction.
+     */
+    public static List<ClusterNode> subtract(Collection<ClusterNode> minuend, Collection<ClusterNode> subtrahend) {
+        return minuend.stream().filter(v -> !subtrahend.contains(v)).collect(Collectors.toList());
+    }
+
+    /**
+     * Adds nodes to the collection of nodes.
+     *
+     * @param op1 First operand.
+     * @param op2 Second operand.
+     * @return Result of the addition.
+     */
+    public static List<ClusterNode> union(Collection<ClusterNode> op1, Collection<ClusterNode> op2) {
```

Comment:
Why do you have collections as parameters?

Reply:
How slow can this be btw? How big are these collections?
I know that "ususally" contains is invoked on a smaller collection. Should we apply such optimization here?